.github/workflows/run-profiling.yaml

```diff
@@ -1,77 +1,78 @@
-# Runs a pyinstrument profiling session on the scale_run profiling model,
-# capturing the profiling result and sending it to the TLOmodel-profiling
-# repository for processing and display.
-#
-# Profiling script executed is src/scripts/profiling/run-profiling.py.
-# Output is the .pyisession file from the profiler, placed into results/
-# results/ folder is then pushed to the TLOmodel-profiling repo, results branch.
-name: Run profiling
+name: Run Profiling
 
 on:
-  workflow_call:
-    inputs:
-      runs-on:
-        description: The type of machine to run job on
-        required: false
-        type: string
-        default: self-hosted
-      event-name:
-        description: |
-          The event name to attribute to the workflow run.
-          Should be passed the value of ${GITHUB_EVENT_NAME} from the triggering workflow.
-        type: string
-        required: false
-        default: workflow_dispatch
-      run-number:
-        description: |
-          The run number (${GITHUB_RUN_NUMBER}) of the triggering workflow.
-          Defaults to 0 if not provided.
-        type: string
-        required: false
-        default: 0
-      sha:
-        description: The SHA to checkout and run profiling on.
-        required: true
-        type: string
-    secrets:
-      profiling-access-token:
-        description: |
-          Access token providing write access to the TLOmodel-profiling repository, 
-          which will be used to commit profiling results on a successful profiling session.
-        required: true
+  workflow_dispatch:
+  # Allow profiling to be triggered by comments on pull requests
+  # Trigger is /run profiling
+  issue_comment:
+    types:
+      - created
+  # Profile the model every Sunday at 00:00, 
+  # on the HEAD of master
+  schedule:
+    - cron: 0 0 * * 6
 
 jobs:
-  run-profiling:
-    name: Profile in developer env
-    runs-on: ${{ inputs.runs-on }}
+  profile-on-comment:
+    name: Comment triggered profiling
+    if: github.event_name == 'issue_comment'
+    uses: ./.github/workflows/run-on-comment.yml
+    with:
+      runs-on: self-hosted
+      keyword: profiling
+      commands: |
+        UNIQUE_ID=${GITHUB_EVENT_NAME}_${GITHUB_RUN_NUMBER}_${GITHUB_SHA}
+        tox -vv -e profile -- --output_name ${UNIQUE_ID}
```

Comment:
To guard against `UNIQUE_ID` becoming out of sync with the corresponding value passed to `--output_name` below then we could potentially have a preceding job which both `profile-on-comment` and `profile-on-dispatch` depend on which sets the identifier and artifact name as outputs and then uses these outputs in the dependent jobs, something like:

```YAML
  set-variables:
    name: Create unique identifier and artifact name for profiling outputs
    runs-on: ubuntu-latest
    outputs:
        unique-id: ${{ steps.set-unique-id.outputs.id }}
        artifact-name: ${{ steps.set-artifact-name.outputs.name }}
    steps:
      - id: set-unique-id
        name: Set identifier output
        run: echo "id=${GITHUB_EVENT_NAME}_${GITHUB_RUN_NUMBER}_${GITHUB_SHA}" >> "$GITHUB_OUTPUT"
      - id: set-artifact-name
        name: Set artifact name output
        run: echo "name=profiling_results_${GITHUB_RUN_NUMBER}" >> "$GITHUB_OUTPUT"

  profile-on-comment:
    needs: set-variables
    ...
    with:
      ...
      commands: tox -vv -e profile -- --output_name ${{ needs.set-variables.unique-id }}
      ...
      artifact-name: ${{ needs.set-variables.artifact-name }}
      ...
 
  profile-on-dispatch:
    needs: set-variables
    ...
    steps:
      ...
      - name: Run profiling in dev environment
        run: tox -vv -e profile -- --output_name ${{ needs.set-variables.unique-id }}
      ...
      - name: Save results as artifact
        uses: actions/upload-artifact@v3
        with:
          name: ${{ needs.set-variables.artifact-name }}
          ...

  upload-profiling-results:
    ...
    needs: [set-variables, profile-on-comment, profile-on-dispatch]
    ...
    steps:
      - name: Download the profiling results
        uses: actions/download-artifact@v3
        with:
          name: ${{ needs.set-variables.artifact-name }}
          path: .
      ...
```

Having written that all out it possibly seems a bit too complex for a minimal avoidance of repetition 😅 so probably not worthwhile!

Reply:
I think we should have the `set-variables` step regardless - I'm not sure how GitHub handles evaluation of variables like ${GITHUB_RUN_NUMBER} (I presume in the sensible way) but this at least means we don't have to worry about it.

It's only adding ~7 extra lines and it does mean there's one fewer place to make a typo in the file names we want to produce!